wagtail/admin/panels.py

```diff
@@ -340,6 +342,12 @@ def is_shown(self):
             """
             return True
 
+        def is_shown_immediately(self):
+            """
+            Whether this panel should be shown immediately (e.g. instead of shown in a modal).
+            """
+            return self.is_shown()
+
```

Comment:
@gasman Could I have an opinion on this method? I haven't found a better way to handle this case:

We want to render the `PublishingPanel` and `CommentingPanel`, but we want to tell `TabbedInterface` to not show the Settings tab if the tab only contains these panels (because they're visually outside the tab).

Reply:
Hi @laymonage,
I think we should investigate whether `PublishingPanel` and `CommentingPanel` need to be panels at all. As I see it, these components are basically hijacking the panels system as a way of injecting extra form fields (or formsets) into WagtailAdminModelForm / WagtailAdminPageForm, but then they have to pretend to be part of the visual panel layout when they really aren't. Up to now we've been able to keep up the fakery and comply with the panels API sufficiently well, but now that we find ourselves introducing a flag effectively saying "this is a fake panel", so that the rest of the framework has to compensate for it, I think that's a sign that the hack has reached breaking point.

I can see two possible approaches here (which might end up being quite similar in practice):

* Rather than giving the top-level panel the job of constructing the form class (via `Panel.get_form_class`), let that happen in the view - i.e. the view code will call the top-level panel's `get_form_options`, merge in whatever additional options we want that _don't_ come from panels (including publishing and commenting - initially these could just be hard-coded in the view but we'd probably want some sort of pluggable system eventually) and call get_form_for_model with the results
* Decouple the commenting / publishing logic from WagtailAdminModelForm / WagtailAdminPageForm, and give them their own forms - so the WagtailAdminModelForm instance is still handling the form fields that are defined through panels, but commenting and publishing are separate form objects that are built in the view code, and validated / processed alongside the main form.

Reply:
Hi @gasman, thanks!
I took a stab at your first suggestion in https://github.com/wagtail/wagtail/pull/8901/commits/a18e14a6a6d9d0b6f0f7161b68547c6574f7058b. Does that look like something you're thinking about?

I can clean up the PR to get rid of the `is_shown_immediately`/`is_custom_rendered` commits if we want to proceed with this approach (or we can just squash later).

Reply:
I've been going back and forth trying to figure out the best approach here, but I'm not sure if we're going to get anything substantially better than your original approach. (Sorry about that!)

My thinking is:
* These things are still basically performing the same job as a panel (namely, modifying the form definition and then rendering some HTML), so it would be overkill to invent a new API just for them
* If we hard-code them into the view rather than making them part of the user-defined panels tree, we need to account for the possibility of users having existing references to these panels in their code, and make sure they don't get defined twice. The neatest way of doing that would be to come up with new class names for the "real" panels to be hard-coded in the view, and make the existing `PublishingPanel` and `CommentingPanel` classes into do-nothing always-hidden panels (so that they have no effect if user code includes them). I thought about coming up with a new name for the concept of panels that aren't part of the visual rendering, like "Utility panels", and renaming them to match (e.g. PublishingUtilityPanel, or moving them to a separate `wagtail.admin.panels.utility` module). But then it feels like we're back to inventing creative naming schemes to obscure the fact that these are "panels that aren't really panels", and we aren't a million miles away from your original approach of calling them "custom rendered". But at least we're no longer giving them fake positions in the panel tree, which counts for something I guess...
* Also - is there a possibility that people might be deliberately omitting these panels for selected page models, as a way of disabling commenting or scheduled publishing - in which case hard-coding them is going to break that pattern? Is that even something we want to support? (https://docs.wagtail.org/en/stable/releases/2.13.html#commentpanel seems to imply that commenting is opt-in, but I doubt that omitting it has been properly tested.) Maybe the best thing to do is not worry about this, and if the requirement does come up later, build a proper system for enabling/disabling them, rather it being emergent behaviour from omitting the panels.

Weighing all of that up, I think it's worth continuing down this route, even if it's only marginally nicer than what you had previously. So that just means tying up the loose end around duplicated panel definitions (i.e. renaming PublishingPanel / CommentingPanel and making the old ones into always-hidden).

Reply:
While working on the rename and anticipating the duplicate panels issue, I just realised this approach actually has another issue:

- To render the fields outside of the edit handler but still make them part of the `Form` instance for validation and stuff, we need to
  - Call `panel.bind_to_model()`
  - Call `panel.get_form_options()` and merge the options with the options from the edit handler, so that the constructed `Form` class includes the fields for the panel
  - Call `panel.get_bound_panel()` to get the `BoundPanel` instance that's ready to be rendered
  - Render them separately by calling `{{ edit_handler.render_form_content }}`, `{{ publishing_panel.render_html }}`, and `{{ comment_panel.render_html }}`.
- However, `edit_handler.render_form_content` uses `edit_handler.render_missing_fields`:
  https://github.com/wagtail/wagtail/blob/44d954e00a6f894eb6951f9fcbb9285a0bf1946b/wagtail/admin/panels.py#L370-L392
- The `render_missing_fields` method scans the form options for fields that aren't part of the edit handler tree, and render them as "ugly unstyled / label-less fields outside of the panel furniture"
  - I didn't notice this problem earlier because without renaming the panels, I didn't think it would be necessary to remove the panels from `set_default_page_edit_handler`. Instead, I had overridden `is_shown()` on the panels to return `False`. This means that the panels are still in the edit handler tree (thus not rendered by `render_missing_fields`), but it's explicitly not rendered as part of the tree.

Now, the problem depends on our answer to this question: "Do we want to include the panels in the default edit handler (even though it's not rendered by the edit handler)?"
- If the answer is yes, then the issue is: it's no longer possible to disable scheduled publishing/commenting by omitting the panels. If users omit the panels (e.g. by setting `settings_panels = []`), then the "ugly fields" will show up (because the panels are no longer in the edit handler tree, but the fields exist in the form options) as duplicates. The features will also still work because they're rendered separately from the panels. There's actually a test for this, albeit a very simple one:
  https://github.com/wagtail/wagtail/blob/44d954e00a6f894eb6951f9fcbb9285a0bf1946b/wagtail/admin/tests/test_edit_handlers.py#L1337-L1350
- If the answer is no, then the issue is: the "ugly fields" will always show up, and we'll have duplicate fields... unless we want to create separate forms.

Both answers come with problems.

There are ways to get around this e.g. by creating separate forms, but I think that'd add even more complexity and I won't be surprised if that introduces new issues.

Honestly, as unfortunate as it is, I think the cleanest, quick approach would be the `is_custom_rendered` flag (or `is_shown_immediately`, or whatever the name should be).
- It allows the features to be toggle-able by including/omitting the panels.
- It doesn't cause issues with "ugly fields" rendered by `render_missing_fields` because the panels are always in the edit handler tree if and only if they're needed
- There will be no duplicate markups because the fields are still rendered by the edit handler.
- It also doesn't mess with the views, which is nice.

Reply:
I suspect that `render_missing_fields` has been redundant for a long time and should be removed, but probably not safe to do that in a minor release.

Let's go with the `is_custom_rendered` approach then. (Would something like `show_panel_furniture` work as a name?)

Reply:
Thanks! Yep, I think that works. Let me adjust the PR again then.

Reply:
Alright, I've changed it to the first approach and renamed the method to `show_panel_furniture`. I've also made it so that the schedule publishing dialog toggle is hidden if the `PublishingPanel` is not present in the edit handler tree (but the scheduled publishing info will still be shown). Tests have also been added for this use case (not so extensive tests though as I don't want to drag this on for too long).